design/keycloak.x/configuration.md

```diff
@@ -0,0 +1,159 @@
+* Use dot (`.`) for separation
+* All options must be in a category (`<categoryName>.<propertyName>`)
+* Multiple levels of categories are permitted (`<categoryName>.<subCategoryName>.<propertyName>`)
+
+Rules for converting property keys to different formats follow these rules:
+
+* Environment variables: 
+  - Prefix with `KC_`
+  - Characters converted to uppercase
+  - Dot (`.`) replaced with underscore (`_`)
+  - Use hyphen (`-`) for camel-case names (`PROPERTY-NAME`)
+* Command-line arguments:
+  - Dot (`.`) converted to hyphen (`-`)
+  
+## SPIs and Provider Configuration
+
+Properties for providers follow the following format:
+
+    spi.<spi-name>.<provider-name>.<config-name>
+
+For example:
+
+    spi.hostname.default.frontendUrl=https://mykeycloak
+    
+## Alias
+
+It should not be required to use SPI/provider specific property names for commonly configured properties for default
+providers. Here we will support alias for properties.
+
+For example `hostname.frontendUrl` is an alias to `spi.hostname.default.frontendUrl`.
+
+## Quarkus options
+
+We do not want to expose users to low-level Quarkus properties, and rather define our own Keycloak specific properties
+when needed.
+
+Depending on the use-case the strategy here will be one of:
+
+* High-level property - for example `db.vendor=oracle12` will behind the scenes set `quarkus.datasource.jdbc.driver` and
+  `quarkus.hibernate-orm.dialect`
+* Mapping/alias - for example `http.ssl.certificate.file` will map to `quarkus.http.ssl.certificate.file`
+
+We will also allow setting Quarkus properties directly. This will not be recommended approach, but available as an option
+to power-users.
+
+## Documenting Properties
+
+We want to document all all properties that a user can set (this does not include the low-level Quarkus properties). We
+would like to achieve something similar to https://quarkus.io/guides/all-config.
+
+In the documentation the property key used in the properties file should be used. It should include a description, the 
+type and the default value.
+
+We would also like the ability to list properties on the command-line by running `bin/kc.sh --list-config-options`.
+
+For SPIs and Providers we should extend the ProviderFactory to add an option method to return information on what
+properties it supports.
+
+## Config Options
+
+The aim is to start small with what config options we provide, and extend when there is a need. Great care should be 
+taken in reducing the amount of configuration needed, where opinionated/best-practice behaviour should be favoured, with
+high-level and simple configuration options introduced only when needed.
+
+Below is a list of config by category we should introduce initially:
+
+### HTTP
+
+Keycloak.X will not come with the capability to generate a self-signed certificate. Instead, Keycloak by default will
+complain if a certificate has not been configured, but will allow explicitly disabling SSL for development purposes.
```

Comment:
This might be understood as suggesting plain HTTP is only OK for development. However, setups with reverse proxy and/or sidecar might also be valid for production.

Reply:
See the section on proxy mode